src/cloud-search-index-storage.js

```diff
@@ -0,0 +1,211 @@
+   * @returns the raw result of the query; will contain the record if it exists.
+   */
+  async get(objectID) {
+    return this.#index.search(
+      '',
+      {
+        facetFilters: [
+          `objectID:${objectID}`,
+        ],
+      },
+    );
+  }
+
+  /**
+   * Save the information in the provided object to the search index as an index record.
+   *
+   * Not all fields provided in the object will be saved to the index. This method takes the values
+   * from the provided object that are specified to be used in the index and saves a record with
+   * those values; other values are ignored.
+   * @param {Object} doc the object to store in the index.
+   * @returns An object containing the saved record's ObjectID.
+   */
+  async save(doc) {
+    const indexRecord = Object.keys(doc).reduce((obj, key) => {
+      const result = clone(obj);
+      if (this.#FIELDS_TO_INDEX.has(key)) {
+        result[key] = doc[key];
+      }
+      return result;
+    }, {});
+    if (!indexRecord?.objectID) {
+      indexRecord.objectID = crypto.randomUUID();
+    }
+
+    this.#logger.info(`Saving doc to cloud record storage index ${this.#indexName}`, indexRecord);
+    return this.#index.saveObject(indexRecord);
+  }
+
+  /**
+   * Get a list of ObjectID that contain the contentHash
+   * @param {String} contentHash sha256 hash of the source binary
+   * @returns list of strings of ojectIDs
+   */
+  async getObjectIdsByContentHash(contentHash) {
+    const searchResult = await this.#index.search(
+      '',
+      {
+        distinct: false,
+        facetFilters: [
+          `contentHash:${contentHash}`,
+        ],
+      },
+    );
+    if (searchResult?.nbHits > 0) {
+      return searchResult?.hits.map((hit) => hit.objectID);
+    }
+    return [];
+  }
+
+  /**
+   * Update all records in the index containing the contentHash
+   *
+   * - Only update the values that are provided in the object (partial update)
+   * - Only add fields that are able to be indexed: `this.#FIELDS_TO_INDEX`
+   * @param {Object} doc document containing the contentHash and other fields to update
+   */
+  // eslint-disable-next-line consistent-return
+  async updateByContentHash(doc) {
+    if (!doc?.contentHash) {
+      const message = 'Missing contentHash, no update in the record storage.';
+      this.#logger.error(message);
+      throw new Error(message);
+    }
+    let objectIDs = await this.getObjectIdsByContentHash(doc.contentHash);
+    if (objectIDs.length > 0) {
+      objectIDs = objectIDs.map((objectID) => ({ objectID, ...doc }));
+      return this.update(objectIDs);
+    }
+    this.#logger.info('No matching record found in the record storage.');
+  }
+
+  /**
+   * Update records in a index
+   * Only update the values that are provided in the object (partial update):
+   * https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/
+   * @param {Array<Object>} records Array of content records with fields to update
+   * @returns
```

Comment:
Do we expect consumers to inspect the result of this? If not, I wouldn't actually return it.

Reply:
the api has a response with a taskId: https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/#response
assuming if there are some timing conflicts or if you wanted to wait() the task until its live in the index, that could be used